File: src/utils/merge_duplicate_exons.m
function genes = merge_duplicate_exons(genes, CFG)
% genes = merge_duplicate_exons(genes, CFG)

if nargin < 2,
    CFG.verbose = 0;
end;

num_removed = 0 ;
for i = 1:length(genes),
    if CFG.verbose && mod(i, 100) == 0, 
        fprintf('%i\r', i);
    end ;

    %%% check if there are non unique exons
    if size(unique(genes(i).splicegraph{1}', 'rows'), 1) == size(genes(i).splicegraph{1}, 2),
        continue;
    end;

    [dummy, exon_order] = sort(genes(i).splicegraph{1}(1, :), 2, 'ascend');
    genes(i).splicegraph{1} = genes(i).splicegraph{1}(:, exon_order);
    genes(i).splicegraph{2} = genes(i).splicegraph{2}(exon_order, exon_order);
    if length(genes(i).splicegraph) > 2,
        genes(i).splicegraph{3} = genes(i).splicegraph{3}(:, exon_order);
    end ;
    exons = genes(i).splicegraph{1} ;
    admat = genes(i).splicegraph{2} ;
    if length(genes(i).splicegraph) > 2,
        initial = genes(i).splicegraph{3}(1, :) ;
        terminal = genes(i).splicegraph{3}(2, :) ;
    end ;
    remove = zeros(1, size(exons, 2)) ;
    for j = 1:size(exons, 2),
        if remove(j), continue ; end ;
        idx = find(exons(1, j) == exons(1, :) & exons(2, j) == exons(2, :)) ;
        idx = setdiff(idx, j) ;
        remove(idx) = 1 ;
        for k=idx,
            if length(genes(i).splicegraph) > 2,
                initial(j) = initial(j) | initial(k) ;
                terminal(j) = terminal(j) | terminal(k) ;
            end;
            admat(j,:) = admat(j,:) | admat(k,:) ;
            admat(:,j) = admat(:,j) | admat(:,k) ;
        end ;
    end ;
    if sum(remove) == 0
        continue ;
    end ;
    genes(i).splicegraph{2} = admat ;
    if length(genes(i).splicegraph) > 2,
        genes(i).splicegraph{3}(1,:) = initial ;
        genes(i).splicegraph{3}(2,:) = terminal ;
    end ;

    genes(i).splicegraph{1} = genes(i).splicegraph{1}(:,find(~remove));
    genes(i).splicegraph{2} = genes(i).splicegraph{2}(find(~remove),find(~remove));
    if length(genes(i).splicegraph) > 2,
        genes(i).splicegraph{3} = genes(i).splicegraph{3}(:, find(~remove));
    end ;
    num_removed = num_removed + sum(remove) ;
end ;
if CFG.verbose,
    fprintf('\n... removed %i duplicate exons ...\n', num_removed);
end;

File: src/utils/merge_duplicate_exons.py
def merge_duplicate_exons(genes, CFG):
# genes = merge_duplicate_exons(genes, CFG)

    num_removed = 0
    for i in range(genes.shape[0]):
        if CFG['verbose'] and i % 100 == 0: 
            print '%i\r' % i

        ### check if there are non unique exons
        if uniqe_rows(genes[i].splicegraph.vertices.T).shape[0] == genes[i].splicegraph.shape[1]:
            continue

        exon_order = sp.argsort(genes[i].splicegraph.vertices[0, :])
        genes[i].splicegraph.vertices = genes[i].splicegraph.vertices[:, exon_order]
        genes[i].splicegraph.edges = genes[i].splicegraph.edges[exon_order, :][:, exon_order]
        genes[i].splicegraph.terminals = genes[i].splicegraph.terminals[:, exon_order]

        exons = genes[i].splicegraph.vertices.copy()
        admat = genes[i].splicegraph.edges.copy()
        initial = genes[i].splicegraph.terminals[0, :]
        terminal = genes[i].splicegraph.terminals[1, :]

        remove = sp.zeros((1, exons.shape[1]), dtype='bool')
        for j in range(exons.shape[1]):
            if remove[j]:
                continue
            idx = sp.where((exons[0, j] == exons[0, :]) & (exons[1, j] == exons[1, :]))[0]
            idx = sp.where(~sp.in1d(idx, j))[0]
            if idx.shape[0] > 0:
                remove[idx] = True
                for k in idx:
                    initial[j] = (initial[j] | initial[k])
                    terminal[j] = (terminal[j] | terminal[k])
                    admat[j, :] = (admat[j, :] | admat[k, :])
                    admat[:, j] = (admat[:, j] | admat[:, k])

        if sp.sum(remove) == 0:
            continue

        genes[i].splicegraph.edges = admat.copy()
        genes[i].splicegraph.terminals[0, :] = initial
        genes[i].splicegraph.terminals[1, :] = terminal

        k_idx = sp.where(~remove)[0]
        genes[i].splicegraph.vertices = genes[i].splicegraph.vertices[:, k_idx]
        genes[i].splicegraph.edges = genes[i].splicegraph.edges[kidx, :][:, k_idx]
        genes[i].splicegraph.terminals = genes[i].splicegraph.terminals[:, k_idx]
        num_removed += sum(remove)

    if CFG['verbose']:
        print '\n... removed %i duplicate exons ...' % num_removed
